### ISSUE TITLE
[RFC] standardize reporting of async errors

### ISSUE THREAD

**Issue by cmyr** [google/xi-editor#543]
# Standardize reporting of errors to client
tl;dr: _Let's use [Failure](https://github.com/withoutboats/failure) + a new `error` RPC notification to report relevant errors to the client._

Error handling in xi-core is currently ad-hoc at best. We would like to propose a standard core-client RPC notification for reporting core errors to the client.

Earlier issues: #434, #438, #519, #484 

## Motivation

Although there is a tentative mechanism for returning errors during RPC _requests_, there is currently no way to report an error that occurs asyncronously.

Such errors could be fatal or recoverable, but in both cases the client should (probably) be notified. For fatal errors the client should be notified so that it can perform any necessary cleanup, write log files, notify the user, restart the connection, or exit cleanly; and with recoverable errors so that it can notify the user, solicit actions, and respond appropriately.

### Recoverable Errors Example

As of [#538](https://github.com/google/xi-editor/pull/538), xi-core will watch open files for external changes. If a file is modified by another process, xi-core will check if the user has modified the local copy. If they have, a request to save the file will fail, because we want to be very defensive about possibly overwriting user data.

In theory, we could introduce a mechanism to present a multi-option modal dialog to the user (#438); but the solution I would prefer is to report this error back to the client, and allow the client to determine the appropriate response.

This allows different clients to handle recoverable errors in the manner most appropriate for their particular needs. In this example case, the client may end up prompting the user for input, but it can do that without the knowledge of xi-core, which saves us from the complications around sync/async response handling required by a `dialog` RPC.

### Portocol

This change involes a new core -> client RPC method, which would look something like this:

```javascript
{
  "method": "error",
  "params": {
    "view_id": "view-id-1", // optional
    "code": 420, // from a canonical list of errors codes
    "msg": "Insufficient chill", // a human readable message.
  }
}
```

Fatal errors could be identified by a reserved block of error codes, or through a separate "fatal_error" method; or we could add a "disconnect" method that core sends when it is shutting down, which could return an exit code. In the long term, xi-core should be very difficult to crash.

### xi-core

What little error reporting currently exists in xi-core is entirely ad-hoc, and involves manually sending the 'alert' rpc method with some custom message. We propose standardizing the handling of errors via the [failure crate](https://github.com/withoutboats/failure)'s [Error + ErrorKind pattern](https://boats.gitlab.io/failure/error-errorkind.html). Our custom `ErrorKind` would enumerate distinct user-expressible errors; for instance an `io::Error` encountered while loading a preference file would be a different `ErrorKind` than an `io::Error` encountered while trying to save a file.

Any method which could fail non-fatally would return a Result<T>; errors will be passed back to the top level handler methods, which will report them to the client automatically.

## Questions

1. How do we want to identify fatal errors? This is actually a concrete problem; core crashes currently hang the runloop without notifying the client, which is less than ideal. This is partly a limitation of the current runloop implementation, our read thread will be blocked on io when our main thread has crashed, but the main thread joins the read thread and doesn't return (#434). While this will be fixed eventually, It probably makes sense for clients to be notified when core has exited abnormally.

**Comment by vlovich** [comment]
Another question is whether we could wrap the underlying RPC errors behind
the same unified error.  Probably not important for front end because, at
least currently, those don’t use rust but if we have requests to the front
end from core core could have a uniform way of handling.  Also, it
simplifies the handling in core if it’s propogaring an RPC error from
somewhere (remote Xi-core, plugins etc) to the front end.

As for fatality, having an is_error function could be used for that.
Alternatively, there could be an is_alive ping the front end could send and
all the processes (core, plugins, remote core etc) would reply
success/error.  Alternatively the core could check the liveliness of
plugins via os APIs and let the front end know about their lifetime.
Similarly the front end could just look for the core’s pix thereby knowing
that it is going to receive notifications about crashes.  Where it can get
tricky potentially is if plugin processes become dynamic in lifetime to
reduce memory/cpu pressure we’ll need a way to indicate graceful shutdown
but I think that’s a tractable problem.

Tldr: my recommendation would be to not worry about trying to send a
message right before process death

**Comment by cmyr** [comment]
Not sure I totally understand the first point; are you talking about a unified Error _type_? We currently have _something_ like this (in [`rust/rpc/src/error.rs`](https://github.com/google/xi-editor/blob/master/rust/rpc/src/error.rs)) but it only really applies to synchronous requests, which require a return value. The overwhelming majority of RPC methods we use are asynchronous, though; the main point of this RFC is to standardize error handling in that case.

I think you're right about the fatal case. Including it in this discussion muddies the waters; it's mostly just a frustration that core hangs instead of exiting after a panic right now, but that's a distinct issue.

**Comment by vlovich** [comment]
I think my point is that explicit error codes as codified in the RFC for
async rpc should be unified with other common error types (Json fails to
decode/encode, RPC failure of some kind etc).  It’s probably an
implementation detail only relevant to rust though.

As for fatal, one thought I had is that Xi-core death must be the easiest
problem since it should result in the IpC channel (TCP?) being terminated
and the front end getting notified.  It’s a harder problem more generally
if you care about the front end getting notified about plug-in deaths (or
remote problems for remote editing)

**Comment by cmyr** [comment]
You're right about the fatal case, in theory; in practice xi-core is bad at actually exiting even when the main thread has crashed, although that is approximately resolved in #546.

There's some subtlety around handling of errors that result from json parsing etc. The JSON-RPC spec does not define any mechanism for communicating errors that occur when handling an async RPC calls ('notifications'), although errors in sync RPC calls ('requests') [are defined](http://www.jsonrpc.org/specification#error_object). I always thought this was an iffy design choice, however. We've _in general_ tried to stay pretty true to the spec, but I'm not opposed to handling RPC-level errors and client level errors through a shared mechanism.

Notifying the client of plugin termination isn't too tricky, since core necessarily notices if either the remote connection closes or the thread crashes, and in both cases can notify the client; this works currently; if you do something that crashes syntect, the checkmark beside it's name in the plugins menu should disappear, although I won't promise this is currently working.
